From a7d7bb54ad30d10370d80dc91adbc329d5a350fa Mon Sep 17 00:00:00 2001
From: Johan Klokkhammer Helsing <johan.helsing@qt.io>
Date: Mon, 29 May 2017 15:44:12 +0200
Subject: [PATCH] Send SurfaceCreated and SurfaceAboutToBeDestroyed events for
 wl_surface

Previously, the events were matching the life cycle of QWaylandWindow. This
commit changes it so the events are matching the life cycle of wl_surfaces
instead (a QWaylandWindow can outlive several wl_surfaces).

Some of these events were already sent in QWindowPrivate (the first and last).
Now we handle the cases where the wl_surface is destroyed and recreated due to
hiding/showing the window or when changing the role of the wl_surface.

Task-number: QTBUG-58423
Change-Id: Ie4a4e7dd529e1a41a2cf42e02cebb3c8aca4a4cc
Reviewed-by: Giulio Camuffo <giulio.camuffo@kdab.com>
Reviewed-by: Marco Martin <notmart@gmail.com>
---
 src/client/qwaylandwindow.cpp | 13 ++++++++++---
 src/client/qwaylandwindow_p.h |  2 +-
 2 files changed, 11 insertions(+), 4 deletions(-)

diff -Naur qtwayland-opensource-src-5.8.0/src/client/qwaylandwindow.cpp qtwayland-opensource-src-5.8.0.tpg/src/client/qwaylandwindow.cpp
--- qtwayland-opensource-src-5.8.0/src/client/qwaylandwindow.cpp	2017-11-04 19:58:01.556672000 +0000
+++ qtwayland-opensource-src-5.8.0.tpg/src/client/qwaylandwindow.cpp	2017-11-04 20:01:35.998664979 +0000
@@ -105,7 +105,7 @@
     delete mWindowDecoration;
 
     if (isInitialized())
-        reset();
+        reset(false);
 
     QList<QWaylandInputDevice *> inputDevices = mDisplay->inputDevices();
     for (int i = 0; i < inputDevices.size(); ++i)
@@ -127,8 +127,11 @@
     if (window()->type() == Qt::Desktop)
         return;
 
-    if (!isInitialized())
+    if (!isInitialized()) {
         initializeWlSurface();
+        QPlatformSurfaceEvent e(QPlatformSurfaceEvent::SurfaceCreated);
+        QGuiApplication::sendEvent(window(), &e);
+    }
 
     if (shouldCreateSubSurface()) {
         Q_ASSERT(!mSubSurfaceWindow);
@@ -222,8 +225,12 @@
     return QPlatformWindow::parent() != Q_NULLPTR;
 }
 
-void QWaylandWindow::reset()
+void QWaylandWindow::reset(bool sendDestroyEvent)
 {
+    if (isInitialized() && sendDestroyEvent) {
+        QPlatformSurfaceEvent e(QPlatformSurfaceEvent::SurfaceAboutToBeDestroyed);
+        QGuiApplication::sendEvent(window(), &e);
+    }
     delete mShellSurface;
     mShellSurface = 0;
     delete mSubSurfaceWindow;
diff -Naur qtwayland-opensource-src-5.8.0/src/client/qwaylandwindow_p.h qtwayland-opensource-src-5.8.0.tpg/src/client/qwaylandwindow_p.h
--- qtwayland-opensource-src-5.8.0/src/client/qwaylandwindow_p.h	2017-11-04 19:58:01.556672000 +0000
+++ qtwayland-opensource-src-5.8.0.tpg/src/client/qwaylandwindow_p.h	2017-11-04 20:02:10.019663827 +0000
@@ -249,7 +249,7 @@
     void initializeWlSurface();
     bool shouldCreateShellSurface() const;
     bool shouldCreateSubSurface() const;
-    void reset();
+    void reset(bool sendDestroyEvent = true);
 
     void handleMouseEventWithDecoration(QWaylandInputDevice *inputDevice, const QWaylandPointerEvent &e);
 
-- 
2.8.3
